lib/internal/modules/run_main.js

```diff
@@ -21,9 +21,6 @@ function resolveMainPath(main) {
 }
 
 function shouldUseESMLoader(mainPath) {
-  const userLoader = getOptionValue('--experimental-loader');
-  if (userLoader)
-    return true;
```

Comment:
Using the `--loader` flag absolutely must enforce the ESM loader, because otherwise the loader will not be applied at all, when it should definitely be used.

Fixing the extension issue on its own should be enough to support the use case of loaders resolving extensionless files though I believe?